abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,203 @@
 """abdiff.core.run_ab_transforms"""
+
+import logging
+import os
+import time
+from datetime import timedelta
+from pathlib import Path
+from time import perf_counter
+
+import docker
+from docker.models.containers import Container
+
+from abdiff.config import Config, configure_logger
+from abdiff.core.exceptions import DockerContainersNotFoundError
+from abdiff.core.utils import read_run_json, update_or_create_run_json
+
+CONFIG = Config()
+logger = logging.getLogger(__name__)
+configure_logger(logger=logger, verbose=False)
+
+# init Docker client
+DockerClient = docker.from_env()
+
+
+def run_ab_transforms(
+    run_directory: str,
+    image_name_a: str,
+    image_name_b: str,
+    input_files: list[str],
+) -> tuple[list[str], ...]:
+    """Run Docker containers with versioned images of Transmogrifier."""
+    start_time = perf_counter()
+
+    # get run ID (used to label containers)
+    run_id = str(run_directory).split("/")[-1]
+
+    # create directories for 'transformed' outputs
+    transformed_directory_a, transformed_directory_b = create_directories(run_directory)
+
+    # create list to track container IDs for a given 'run'
+    container_ids = []
+
+    # create list of run configs
+    run_configs = [
+        (image_name_a, transformed_directory_a),
+        (image_name_b, transformed_directory_b),
+    ]
+
+    # transform input files using images A and B via Docker containers
+    # save transformed record JSON files in corresponding directory
+    for input_file in input_files:
+        source = (input_file.split("/")[-1]).split("-")[0]
+        for docker_image, transformed_directory in run_configs:
+            container = DockerClient.containers.run(
+                docker_image,
+                command=generate_transform_command(source, input_file),
+                detach=True,
+                environment={
+                    "AWS_ACCESS_KEY_ID": CONFIG.AWS_ACCESS_KEY_ID,
+                    "AWS_SECRET_ACCESS_KEY": CONFIG.AWS_SECRET_ACCESS_KEY,
+                    "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+                },
+                labels={
+                    "docker_image": docker_image,
+                    "source": source,
+                    "input_file": input_file,
+                    "run": run_id,
+                },
+                volumes=[f"{transformed_directory}:/tmp"],
+            )
+            logger.info(
+                f"Container '{container.id}' (Docker image: {docker_image}) "
+                f"RUNNING transform for '{source}' input_file: {input_file}."
+            )
```

Comment:
Maybe worth breaking this out into a standalone function? One that kicks back a running container?

Reply:
Let me know what you think of the [`run_docker_container` function](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/6/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcR71-R104). Note the removal of the `generate_transform_command`; after reworking the parsing method, the remaining contents of the `generate_transform_command` felt minimal / not requiring its own function.

Reply:
Agreed!  The `run_docker_container` function looks great; one purpose, easy to follow.  Will wait for the "real" PR to dig into how and when it's called, but doesn't feel like it needs any further refactoring.